server/sheet.go

```diff
@@ -67,6 +72,206 @@ func (s *Server) registerSheetRoutes(g *echo.Group) {
+			}
+
+			lastCommit, err := vcsPlugin.Get(vcs.Type, vcsPlugin.ProviderConfig{Logger: s.l}).FetchCommitByID(ctx,
+				common.OauthContext{
+					ClientID:     vcs.ApplicationID,
+					ClientSecret: vcs.Secret,
+					AccessToken:  repo.AccessToken,
+					RefreshToken: repo.RefreshToken,
+					Refresher:    s.refreshToken(ctx, repo.ID),
+				},
+				vcs.InstanceURL,
+				repo.ExternalID,
+				fileMeta.LastCommitID,
+			)
+			if err != nil {
+				return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Failed to fetch commit data from VCS, instance URL: %s, repoId: %s, last commit ID: %s", vcs.InstanceURL, repo.ExternalID, fileMeta.LastCommitID)).SetInternal(err)
+			}
+
+			sheetVCSPayload := &api.SheetVCSPayload{
+				FileName:     fileMeta.Name,
+				FilePath:     fileMeta.Path,
+				Size:         fileMeta.Size,
+				Author:       lastCommit.AuthorName,
+				LastCommitID: lastCommit.ID,
+				LastSyncTs:   time.Now().Unix(),
+			}
+			payload, err := json.Marshal(sheetVCSPayload)
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, "Failed to marshal sheetVCSPayload").SetInternal(err)
+			}
+
+			payloadString := string(payload)
+			vscSheetType := api.SheetForSQL
+			sheet, err := s.SheetService.FindSheet(ctx, &api.SheetFind{
+				Name:      &sheetInfo.SheetName,
+				ProjectID: &project.ID,
+				Source:    (*api.SheetSource)(&vcs.Type),
+				Type:      (*api.SheetType)(&vscSheetType),
+			})
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to find sheet with name: %s, projectID: %d", sheetInfo.SheetName, projectID)).SetInternal(err)
+			}
+
+			if sheet == nil {
+				sheetCreate := api.SheetCreate{
+					ProjectID:  projectID,
+					CreatorID:  c.Get(getPrincipalIDContextKey()).(int),
+					Name:       sheetInfo.SheetName,
+					Statement:  fileContent,
+					Visibility: api.PublicSheet,
+					Source:     api.SheetFromBytebase,
+					Type:       api.SheetForSQL,
+					Payload:    payloadString,
+				}
+
+				// In non-tenant mode, we can set a databaseId for sheet.
+				if project.TenantMode != api.TenantModeDisabled {
+					if sheetInfo.EnvironmentName != "" && sheetInfo.DatabaseName != "" {
+						databaseList, err := s.composeDatabaseListByFind(ctx, &api.DatabaseFind{
+							Name:      &sheetInfo.DatabaseName,
+							ProjectID: &projectID,
+						})
+						if err != nil {
+							return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to find database list with name: %s, projectID: %d", sheetInfo.DatabaseName, projectID)).SetInternal(err)
+						}
+
+						for _, database := range databaseList {
+							if sheetCreate.DatabaseID != nil {
+								break
+							}
+
+							labelList := []*api.DatabaseLabel{}
+							err := json.Unmarshal([]byte(database.Labels), &labelList)
```

Comment:
database always belong to an instance which in turns belong to an environment. So it's more common to access via database.instance.environment

Accessing Label is OK, since the environment label derives from the environment as well. It's just not the typical way to retrieve environment via label.